### PR TITLE
Notes on the latest CI in SageMath

### PR DESCRIPTION
**Pull request by saraedum**
@embray, I wrote a bit over a page, I hope that's not too much. Feel free to modify this in any way that you see fit.

### CODE REVIEW

**Comment by saraedum** [comment]
Also, I do not distinguish clearly between what's already merged and what is still pending. Almost everything I am talking about is "needs review", so I guess that's fine. We could probably mention at some point that not all of this is available to all developers yet.

**Comment by embray** [comment]
> Also, I do not distinguish clearly between what's already merged and what is still pending.

That's fine.  Same goes for my report for D3.5--most of the work is "done" but it's not yet merged; live yet.  If I notice anything that seems unrealistic I'll adjust it.  "A bit over a page" is perfect, thanks.

File: WP3/D3.8/report.tex [hunk 1]
```diff
@@ -185,13 +185,94 @@ \section{Continuous integration achievements of OpenDreamKit projects}\label{pro
...
-% leave it to the conclusion.
+The SageMath project had traditionally employed two distinct CI systems, the
+Buildbot and Sage's own development, the
+patchbot\footnote{\url{https://github.com/sagemath/sage-patchbot}}.
+
+\subsubsection{Buildbot and patchbot}
+Sage's Buildbot runs as a non-advisory CI system on a wide variety of
+platforms. It builds Sage and runs its test suite for every change that gets
+accepted into the mainline of the Sage code. Its main purpose is not to be an
+advisory tool for developers but to catch problems and incompatibilities of
```
Comment (embray) [inline WP3/D3.8/report.tex]: Of course, the way it is used currently, in combination with Volker, it does kind of serve an advisory role since he is basically manually merging changes, testing them, and the backing them out if they don't integrate.  But I think maybe that's a fine point we don't want to go into detail on here.

If I were him (and if we had a better advisory CI system) I would just merge changes directly into master, and in the relatively rare case that something breaks on one platform or another I would open a critical issue to fix it (assuming it's a blocker, on a platform we care about).

I don't necessarily think one way is "better" than the other, but his process seems much more labor intensive.  I can abide a broken build as long as it's quickly fixed.  He cannot abide ever having a broken build, it seems :|

File: WP3/D3.8/report.tex [hunk 2]
```diff
@@ -185,13 +185,94 @@ \section{Continuous integration achievements of OpenDreamKit projects}\label{pro
...
+Sage or its dependencies, in particular on more obscure platforms that
+dependent projects do often not test for. Due to the vast range of platforms
+covered, the Buildbot can not realistically be turned into an advisory CI
+system as it is not conceivable to provision and administer the number of
+machines that would be necessary to run Sage's test suite at every stage of a
+test that gets proposed by a developer.
```
Comment (embray) [inline WP3/D3.8/report.tex]: I find the wording of the end of this sentence "run Sages test suite at every stage of a test that gets proposed by a developer".  I'm pretty sure I know what you mean here; I can try to think of a way to reword it

File: WP3/D3.8/report.tex [hunk 3]
```diff
@@ -185,13 +185,94 @@ \section{Continuous integration achievements of OpenDreamKit projects}\label{pro
...
+
+Therefore, the Sage community at some point developed its own patchbot system
```
Comment (embray) [inline WP3/D3.8/report.tex]: Here I would put "patchbot" in quotes (or maybe even Title Case) since it's a unique name of a product we developed.

File: WP3/D3.8/report.tex [hunk 4]
```diff
@@ -185,13 +185,94 @@ \section{Continuous integration achievements of OpenDreamKit projects}\label{pro
...
+which runs Sage's test suite at every stage of a proposed change. Anybody can
+run a patchbot on their private computer, typically a Linux system. These
+patchbots will then look for proposed changes to Sage, run Sage's test suite,
+and report findings back to a central server. It is worth noting that the
+patchbot system predates most of the currently popular CI systems.
+
+The patchbot system comes with a number of issues: It is not nearly as polished
+as the professional offerings which means that its workings can often be
```
Comment (embray) [inline WP3/D3.8/report.tex]: I have a problem with the use of "professional offerings" in this case.  While it's certainly true that the patchbot is unpolished, I wouldn't consider it "unprofessional" either which this sort of implies by contrast.

I'm not sure what to suggest instead though.  Maybe "commercial offerings" would be valid, though it should be clear that "commercial" does not mean we have to pay for them (as all these services are free for open source projects--perhaps something I should clarify earlier in the report).

File: WP3/D3.8/report.tex [hunk 5]
```diff
@@ -185,13 +185,94 @@ \section{Continuous integration achievements of OpenDreamKit projects}\label{pro
...
+confusing and that it also requires continuous maintenance and development from
+the SageMath community and by those that run the patchbots on their private
+machines. The fact that it is running on private machines means that due to
+load, connectivity issues, and other temporary issues on these machines, the
+patchbots are very frequently reporting false positives. This noise can be very
+frustrating as it makes the patchbot system unreliable and in particular harder
+to understand for new contributors. Finally, due to limitations of the patchbot
+architecture, not all types of changes can be tested, in particular it can not
```
Comment (embray) [inline WP3/D3.8/report.tex]: "can be tested," -> "can be tested;"

File: WP3/D3.8/report.tex [hunk 6]
```diff
@@ -185,13 +185,94 @@ \section{Continuous integration achievements of OpenDreamKit projects}\label{pro
...
+test changes that involve dependent software packages. This can be frustrating
+as testing such changes is often the most time consuming task for developers
+and it would be particularly nice to give such tasks away to a CI system.
```
Comment (embray) [inline WP3/D3.8/report.tex]: Also worth noting that it's the kind of change most likely to cause things to break.

I was working a few months ago to an update to the patchbot that would allow it to test such tickets efficiently but I ran into some problems and gave up on it. Not that those problems were insurmountable at all, I just needed to change my focus to other things.  I don't know if it's worth mentioning this--if the new CI continues to work I'll probably prefer to focus my effort on that, over improvements to the patchbot.

File: WP3/D3.8/report.tex [hunk 7]
```diff
@@ -185,13 +185,94 @@ \section{Continuous integration achievements of OpenDreamKit projects}\label{pro
...
+
+\subsubsection{Modern CI with GitLab CI and CircleCI}
+To fix these issues we were looking at best pratices that would give the
+SageMath community fast, reproducible, reliable CI with minimal maintenance and
+continuous development necessary to keep the CI working for everybody.  We
+settled for GitLab CI, a modern open-source CI system, and CircleCI, a popular
+closed-source system. Which one to use is up to each developer. For simplicity
+we focus on GitLab CI here. GitLab CI typically uses docker containers as its
```
Comment (embray) [inline WP3/D3.8/report.tex]: Perhaps worth stating explicitly that most of what we write about GitLab here applies fairly similarly to Circle CI.  Also I would be interested in a note about why you chose to try two systems in the first place, and whether we intend to maintain both systems long term, or if it's more a question of evaluating them both and then focusing on whichever one works better.

Comment (embray) [reply WP3/D3.8/report.tex]: I can just fix this during final editing, but we should write "Docker" not "docker".

File: WP3/D3.8/report.tex [hunk 8]
```diff
@@ -185,13 +185,94 @@ \section{Continuous integration achievements of OpenDreamKit projects}\label{pro
...
+backend (but it can also be based on many other technologies.) These containers
+are essentially virtual machines that can be hosted in a central location but
+also on private computers. Unlike in the patchbot systems, docker containers
+are standardized and isolated so that there is usually no issue of false
+positives. % note that I do not distinguish between docker containers and
+% images; I think it's more confusing to the reader to make such a distinction.
```
Comment (embray) [inline WP3/D3.8/report.tex]: I agree, though later on I might say something more about this distinction.  @jdemeyer raised an interesting point that he would like to be able to test Cysignals on more platforms.  And in the case of Cysignals in particular I can imagine cases where one would like to be able to test it against different kernels, which is not possible with containers.

File: WP3/D3.8/report.tex [hunk 9]
```diff
@@ -185,13 +185,94 @@ \section{Continuous integration achievements of OpenDreamKit projects}\label{pro
...
+
+GitLab CI now runs a "pipeline" for every change proposed by a developer.
+\TODO{Insert \url{https://trac.sagemath.org/raw-attachment/ticket/24655/gitlab.png} here.}
+This pipeline goes in stages. In the setup for SageMath, the first stage builds
+several flavors of docker container which contain a built version of Sage. The
+second stage runs tests on these containers. Finally the docker containers from
+the first stage are made publicly available.
```
Comment (embray) [inline WP3/D3.8/report.tex]: It might be nice to add a screenshot of a pipeline here.  I can add that.

File: WP3/D3.8/report.tex [hunk 10]
```diff
@@ -185,13 +185,94 @@ \section{Continuous integration achievements of OpenDreamKit projects}\label{pro
...
+download these containers; in typical cases they only have to download a few MB
+to get a hold of the docker container.
+
+\subsubsection{Use Cases of Docker Containers}
+Having these docker containers readily available as a by-product of our CI turned out to be very valuable.
+\TODO{Maybe insert \url{https://trac.sagemath.org/raw-attachment/ticket/24655/dockerhub.png} here.}
+
+It means that recent stable \& unstable versions of Sage are automatically made
+available for anybody to work with, without the need of having to understand
+how to build Sage on their machine.
```
Comment (embray) [inline WP3/D3.8/report.tex]: The need to understand *or* take the often considerable time :)

File: WP3/D3.8/report.tex [hunk 11]
```diff
@@ -185,13 +185,94 @@ \section{Continuous integration achievements of OpenDreamKit projects}\label{pro
...
+
+Sage developers can use these to reproduce and understand issues
+reported by the CI system. Simply by downloading the container and running the
+Sage contained in that container.
```
Comment (embray) [inline WP3/D3.8/report.tex]: I would reword this.  Something like:

"Sage developers can then use the containers that result from builds to reproduce and understand issues reported by the CI system simply by downloading it and running the build of Sage in that container.  This is far more efficient than attempting to reproduce a build manually on one's local machine.

File: WP3/D3.8/report.tex [hunk 12]
```diff
@@ -185,13 +185,94 @@ \section{Continuous integration achievements of OpenDreamKit projects}\label{pro
...
+
+Projects that depend on Sage can use these docker containers for their own CI
```
Comment (embray) [inline WP3/D3.8/report.tex]: It might be worth noting explicitly here that CI systems have been moving largely towards using containerization technology in their build infrastructure, and thus giving users the ability to supply container images as the basis for their build, rather than relying solely on the default build images provided by the CI system.

**Comment by embray** [comment]
Thank you again, this is otherwise perfect!

File: WP3/D3.8/report.tex (hunk 1 above)
Comment (saraedum) [reply WP3/D3.8/report.tex]: Yes, I was aware that it is technically advisory. For developers it does not feel like an advisory CI, however.
I agree that the current style seems relatively labor intensive though much more automated than what we used to do if I'm informed correctly.

File: WP3/D3.8/report.tex (hunk 2 above)
Comment (saraedum) [reply WP3/D3.8/report.tex]: Yes, I wanted to work around the word "commit" here which feels quite cumbersome. If you have better wording please change it.

File: WP3/D3.8/report.tex (hunk 3 above)
Comment (saraedum) [reply WP3/D3.8/report.tex]: Sounds good to me.

File: WP3/D3.8/report.tex (hunk 4 above)
Comment (saraedum) [reply WP3/D3.8/report.tex]: You're right the wording is unfortunate. Maybe dedicated? The idea is that it's not homegrown...commercial is also valid I think.

File: WP3/D3.8/report.tex (hunk 9 above)
Comment (saraedum) [reply WP3/D3.8/report.tex]: I think I added a TODO that has the URL of such a screenshot.

File: WP3/D3.8/report.tex (hunk 7 above)
Comment (saraedum) [reply WP3/D3.8/report.tex]: GitLab CI did not integrate with GitHub back then. Now it does I heard, so maybe there is not that much of a point anymore. You can also SSH into the machines running jobs on CircleCI. GitLab is actively working on that feature but they don't have it yet. Then again that's not so terribly important anymore since I am now basing everything on the docker images so you can just download these and run everything yourself if you want to debug things.
So, we'll probably drop CircleCI support again in the long run.

File: WP3/D3.8/report.tex (hunk 6 above)
Comment (saraedum) [reply WP3/D3.8/report.tex]: I believe there were attempts in that direction before but it's just quite tricky to reset efficiently to a working state without VMs and snapshotting. SPKGs are hopefully going away eventually, so we should also not build too much to cater for something that I personally consider obsolete; there is of course no full replacement yet, but a requirements.txt file for conda already gets you quite far on mainstream platforms. If I focused completely on this I would probably be able to make that work for regular developers in my time that remains with logilab. But that's a bit outside the scope of this PR ;)

File: WP3/D3.8/report.tex (hunk 8 above)
Comment (saraedum) [reply WP3/D3.8/report.tex]: True.

File: WP3/D3.8/report.tex (hunk 11 above)
Comment (saraedum) [reply WP3/D3.8/report.tex]: Sounds good to me.

**Comment by saraedum** [comment]
Ok. It'd be great if you could make the changes that you proposed. It's probably easier if you just take it from here as you are editing in other places as well. But I don't mind making these changes myself either.

**Comment by embray** [comment]
Ok, I'll just do that then.
